source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -0,0 +1,195 @@
+  } else {
+    direction_config.stats().not_decompressed_.inc();
+    ENVOY_STREAM_LOG(debug, "do not decompress {}: {}", callbacks, direction_config.logString(),
+                     headers);
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus DecompressorFilter::maybeDecompress(
+    const DecompressorFilterConfig::DirectionConfig& direction_config,
+    Compression::Decompressor::DecompressorOptRef decompressor,
+    Http::StreamFilterCallbacks& callbacks, Buffer::Instance& input_buffer) const {
+  if (decompressor.has_value()) {
+    Buffer::OwnedImpl output_buffer;
+    decompressor->get().decompress(input_buffer, output_buffer);
+
+    // Report decompression via stats and logging before modifying the input buffer.
+    direction_config.stats().total_compressed_bytes_.add(input_buffer.length());
+    direction_config.stats().total_uncompressed_bytes_.add(output_buffer.length());
+    ENVOY_STREAM_LOG(debug, "{} data decompressed from {} bytes to {} bytes", callbacks,
+                     direction_config.logString(), input_buffer.length(), output_buffer.length());
+
+    input_buffer.drain(input_buffer.length());
+    input_buffer.add(output_buffer);
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+bool DecompressorFilter::hasCacheControlNoTransform(
+    Http::RequestOrResponseHeaderMap& headers) const {
+  return headers.CacheControl()
+             ? StringUtil::caseFindToken(headers.CacheControl()->value().getStringView(), ",",
+                                         Http::Headers::get().CacheControlValues.NoTransform)
+             : false;
+}
+
+bool DecompressorFilter::contentEncodingMatches(Http::RequestOrResponseHeaderMap& headers) const {
+  if (headers.ContentEncoding()) {
+    absl::string_view coding = StringUtil::trim(
+        StringUtil::cropRight(headers.ContentEncoding()->value().getStringView(), ","));
+    if (StringUtil::CaseInsensitiveCompare()(config_->contentEncoding(), coding)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void DecompressorFilter::removeContentEncoding(Http::RequestOrResponseHeaderMap& headers) const {
+  const auto all_codings = headers.ContentEncoding()->value().getStringView();
+  const auto remaining_codings = StringUtil::trim(StringUtil::cropLeft(all_codings, ","));
+
+  if (remaining_codings != all_codings) {
+    headers.setContentEncoding(remaining_codings);
+  } else {
+    headers.removeContentEncoding();
+  }
+}
+
+/**
+ * Add "chunked" to the Transfer-Encoding header if it's not there yet.
+ */
+void DecompressorFilter::sanitizeTransferEncoding(Http::RequestOrResponseHeaderMap& headers) const {
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  if (!(transfer_encoding &&
+        StringUtil::caseFindToken(transfer_encoding->value().getStringView(), ",",
+                                  Http::Headers::get().TransferEncodingValues.Chunked,
+                                  true /* trim_whitespace */))) {
```

Comment:
Do we need to trim whitespace elsewhere in the above header checks too?

Reply:
which one, in particular?

Reply:
`contentEncodingMatches` and `hasCacheControlNoTransform`

Reply:
they do trimming were necessary.